### PR TITLE
Mute track

### PR DESCRIPTION
**Pull request by kamil-stasiak**
## Description

- The `replaceTrack` method can handle a nullable track parameter. A previously renegotiated track can be stopped and reused in the future.
- The `addTrack` method doesn't require the `stream` parameter. The FishajmClient creates a new internal `Stream` object for every `addTrack` invocation, allowing updates to this object. In the previous implementation, a new track was never added to a stream (likely because that stream was an external dependency).
- The `localTrackMuted` event is emitted when a track is replaced with `null`.
- The `localTrackUnmuted` event is emitted when a track is replaced with a non-null track object.

### E2E tests
To test if this works, I've added tests that check if another peer receives frames and if the background color changes.

## Motivation and Context

This change allows disabling a media device when it's not in use without additional renegotiation. Previously, it was necessary to add and remove the track every time the camera or microphone was started or stopped.

## Types of changes

- [ ] Bug fix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [x] Breaking change (a fix or feature that would cause existing functionality to not work as expected)
  - The `replaceTrack` method can handle a nullable track parameter.
  - The `stream` parameter was removed from the `addTrack` method.

### CODE REVIEW

File: e2e/app/src/App.tsx [hunk 1]
```diff
@@ -9,10 +9,10 @@ import {
   BandwidthLimit,
   SimulcastConfig,
 } from '@fishjam-dev/ts-client';
-import { PeerMessage } from './protos/fishjam/peer_notifications';
 import { useEffect, useState, useSyncExternalStore } from 'react';
 import { MockComponent } from './MockComponent';
 import { VideoPlayerWithDetector } from './VideoPlayerWithDetector';
+import { PeerMessage } from '../../../src/protos/fishjam/peer_notifications';
```
Comment (mironiasty) [inline e2e/app/src/App.tsx:15]: Is this correct import? I think that it should be imported via `@fishjam-dev/ts-client` path

File: e2e/app/src/MuteTrackTest.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    };
+
+    webrtc.on('localTrackAdded', localTrackAdded);
+    webrtc.on('localTrackReplaced', localTrackReplaced);
+
+    return () => {
+      webrtc.removeListener('localTrackAdded', localTrackAdded);
+      webrtc.removeListener('localTrackReplaced', localTrackReplaced);
+    };
+  }, []);
```
Comment (mironiasty) [inline e2e/app/src/MuteTrackTest.tsx:41]: Does `webrtc` should be added as dependency to this array?

File: e2e/app/src/VideoPlayerWithDetector.tsx [hunk 3]
```diff
@@ -17,17 +19,53 @@ const rgbToText = (pixel: Pixel): string => {
...
+    }, 50);
+
+    return () => {
+      clearInterval(id);
+    };
+  }, [stream]);
+
   useEffect(() => {
-    const id = setInterval(() => {
+    const id = setInterval(async () => {
```
Comment (mironiasty) [inline e2e/app/src/VideoPlayerWithDetector.tsx:68]: Why `async`?

File: src/FishjamClient.ts [hunk 4]
```diff
@@ -662,8 +672,14 @@ export class FishjamClient<
...
-    this.webrtc?.on('localTrackEncodingBandwidthSet', (event) => {
-      this.emit('localTrackEncodingBandwidthSet', event);
+    this.webrtc?.on('localTrackMuted', (event) => {
+      this.emit('localTrackMuted', event);
+    });
+    this.webrtc?.on('localTrackUnmuted', (event) => {
+      this.emit('localTrackUnmuted', event);
+    });
+    this.webrtc?.on('localTrackBandwidthSet', (event) => {
+      this.emit('localTrackBandwidthSet', event);
```
Comment (mironiasty) [inline src/FishjamClient.ts:682]: This is outside of this PR scope, but I would like to discuss why FishjamClient class has so many functions 😮

File: src/webrtc/webRTCEndpoint.ts [hunk 5]
```diff
@@ -1333,24 +1355,53 @@ export class WebRTCEndpoint<
...
-          }
-        })
-        .finally(() => {
-          this.resolvePreviousCommand();
-          this.ongoingTrackReplacement = false;
-          this.processNextCommand();
-        });
+
+    const track = this.trackIdToSender.get(trackId);
+    const sender: RTCRtpSender | null = track?.sender ?? null;
```
Comment (mironiasty) [inline src/webrtc/webRTCEndpoint.ts:1360]:
```suggestion
    const sender = track?.sender ?? null;
```

File: src/webrtc/webRTCEndpoint.ts [hunk 6]
```diff
@@ -1333,24 +1355,53 @@ export class WebRTCEndpoint<
...
+
+        if (newTrackMetadata) {
+          this.updateTrackMetadata(trackId, newTrackMetadata);
+        }
+      })
+      .finally(() => {
+        this.resolvePreviousCommand();
+        this.ongoingTrackReplacement = false;
+        this.processNextCommand();
+      });
```
Comment (mironiasty) [inline src/webrtc/webRTCEndpoint.ts:1404]:
```suggestion
    try {
      await sender.replaceTrack(newTrack);
      trackContext.track = newTrack;
      if (newTrackMetadata) {
        this.updateTrackMetadata(trackId, newTrackMetadata);
      }
    } catch (err) {
      //ignore
    } finally {
      this.resolvePreviousCommand();
      this.ongoingTrackReplacement = false;
      this.processNextCommand();
    }
```

File: src/FishjamClient.ts (hunk 4 above)
Comment (kamil-stasiak) [reply src/FishjamClient.ts:682]: It is very convenient to use events to synchronize the state between TypeScript and the React client. Emitting events for every action allows us to react to every internal state change in the TypeScript client. This was introduced to simplify synchronization logic and to eliminate some race conditions

File: e2e/app/src/App.tsx (hunk 1 above)
Comment (kamil-stasiak) [reply e2e/app/src/App.tsx]: What do you think about hiding `protos` inside an additional namespace?
```
import { PeerMessage } from '@fishjam-dev/ts-client/protos';
```

Comment (mironiasty) [reply e2e/app/src/App.tsx]: > What do you think about hiding protos inside an additional namespace?

Good idea :)

File: package.json [hunk 7]
```diff
@@ -18,7 +18,8 @@
     "dist/src"
   ],
   "exports": {
-    ".": "./dist/src/index.js"
+    ".": "./dist/src/index.js",
+    "./protos": "./dist/src/protos/index.js"
```
Comment (mironiasty) [inline package.json:22]: 👍

**Review by mironiasty — APPROVED** [review]

**Pull request merged**